Quaternion/tests/test_all.py

```diff
@@ -1,12 +1,60 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 import numpy as np
+import pytest
+
 from .. import Quat
 
 ra = 10.
 dec = 20.
 roll = 30.
 q0 = Quat([ra, dec, roll])
 
+equatorial_23 = np.array([[[ 10,  20,  30],
```

Comment:
For validation that the *vectorized operations* are working, we can stipulate that the scalar code works by seeing that the existing tests pass.  At that point the actual values don't matter so much as confirming that doing things using the new interface with non-scalar inputs gives the same answer as doing through the old scalar interface (on a per-element basis).  So something like (untested!):
```
equatorial_23 = np.linspace(1, 50, 2 * 3 * 3).reshape(2, 3, 3)
q_23 = np.empty(shape=(2, 3, 4))
transform_23 = np.empty(shape=(2, 3, 3, 3))
for ii in range(2):
    for jj in range(3):
        q = Quat(equatorial_23[ii, jj])
        q_23[ii, jj] = q.q
        transform_23[ii, jj] = q.transform
```
This should let you compare the vector interface to the scalar version to machine precision instead of `allclose` (which by default is only 6 digits).  (The plan is that the exact same math gets done at the end of the day).  Since these values are well-behaved you should be able to do round-trip tests and get answers that match to within pretty tight tolerances (one part in 1e-10 or better).

The idea of including the bad corner cases in the vector testing is good, but maybe just do that as a separate one-off.  Could be more minimal, just `[[10, -90, 30], [10, 90, -30]]`.

Reply:
To do these "round-trip" check, I wrote a little generator helper function (maybe something exists already?) so I write something like

```
q1_inv = q1.inv()
for i in indices(q1._shape):
    # check that Quat(q).inv().q[i] == Quat(q[i]).inv().q
    assert np.all(q1_inv.q[i] == Quat(q=q1.q[i]).inv().q)
```

and 

```
q = Quat(equatorial=equatorial_23)
for i in indices(q._shape):
    # check that Quat(equatorial).q[i] == Quat(equatorial[i]).q
    assert np.all(q.q[i] == Quat(equatorial_23[i]).q)
```

Note that in this example I used the (private) _shape data member, so I am thinking of making it public. I do not use _shape in the version I am about to check in the repository.

Reply:
ok, in commit 48c466dbfe5cd I removed these extra arrays.

Let it be noted that I do not like it. In my opinion, it decreases readability, it does not improve testing (the tests already were checking with full precision) and there is non-trivial code running out of tests.

Reply:
Noted.  

For me as the code reviewer reading the original code, the big glop of numbers was meaningless since I have no clue how they were generated or what they mean, and it was difficult to discern by eye the dimensionality of the arrays.  If you included a comment that showed the exact code to generate those values (and e.g. said that you ran it with the  current *release* version of Quaternion) then I might let it pass.

With the programmatic version the I can see instantly what's happening, in particular that the scalar version of Quaternion is being used to generate a multi-dimensional (list of lists) of reference values.

Just a difference in perspective between the person writing vs. reading the code.  :smile: